jetty-webapp/src/main/java/org/eclipse/jetty/webapp/DecoratingListener.java

```diff
@@ -0,0 +1,180 @@
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.webapp;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import javax.servlet.ServletContextAttributeEvent;
+import javax.servlet.ServletContextAttributeListener;
+
+import org.eclipse.jetty.util.Decorator;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A ServletContextAttributeListener that listens for a specific attribute
+ * name (default "org.eclipse.jetty.webapp.Decorator") to obtain a
+ * decorator instance from the webapp.  The instance is then either coerced
+ * to a Decorator or reflected for decorator compatible methods so it can
+ * be added to the {@link WebAppContext#getObjectFactory()} as a
+ * {@link Decorator}.
+ */
+public class DecoratingListener implements ServletContextAttributeListener
+{
+    public static final String DECORATOR_ATTRIBUTE = "org.eclipse.jetty.webapp.decorator";
+    private static final Logger LOG = Log.getLogger(DecoratingListener.class);
+    private static final MethodType decorateType;
+    private static final MethodType destroyType;
+
+    static
+    {
+        try
+        {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            decorateType = MethodType.methodType(Object.class, Object.class);
+            destroyType = MethodType.methodType(Void.TYPE, Object.class);
+            // Ensure we have a match
+            lookup.findVirtual(Decorator.class, "decorate", decorateType);
+            lookup.findVirtual(Decorator.class, "destroy", destroyType);
+        }
+        catch (Exception e)
+        {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private final WebAppContext _context;
+    private final String _attributeName;
+    private Decorator _decorator;
+
+    public DecoratingListener()
+    {
+        this(null, null);
+    }
+
+    public DecoratingListener(String attributeName)
+    {
+        this(null, attributeName);
+    }
+
+    public DecoratingListener(WebAppContext context)
+    {
+        this(context, null);
+    }
+
+    public DecoratingListener(WebAppContext context, String attributeName)
+    {
+        _context = context == null ? WebAppContext.getCurrentWebAppContext() : context;
+        _attributeName = attributeName == null ? DECORATOR_ATTRIBUTE : attributeName;
+        if (_context.getAttribute(getClass().getName()) == null)
+            _context.setAttribute(getClass().getName(), attributeName);
```

Comment:
I fail to understand this.
Using `getClass().getName()` as a key is intended? Using the local variable `attributeName` instead of the field `_attributeName` is intended?

Using `getClass().getName()` may be a problem if this class is subclassed, as you expect a certain key (the base class) but you get another (the sub class).

IIUC this is the attribute key that defines what is the other attribute key whose value is the decorator object.
Should not this key be called `DECORATOR_ATTRIBUTE_NAME_ATTRIBUTE` and be a constant?
 It's a mouthful but it's clear what it is.

A comment about why this attribute is set will help.